### PR TITLE
Leaks during client authentication

### PR DESCRIPTION
**Pull request by tburgin**
This only seems to happen during client authentication:

```
Leaked Object   #   Address Size    Responsible Library Responsible Frame
Malloc 64 Bytes 2   < multiple >    128 Bytes   Security    DefaultAllocator::malloc(unsigned long)
Malloc 64 Bytes 1   0x62000007b8c0  64 Bytes    Security    DefaultAllocator::malloc(unsigned long)
Malloc 16 Bytes 2   < multiple >    32 Bytes    Security    DefaultAllocator::malloc(unsigned long)
Malloc 64 Bytes 2   < multiple >    128 Bytes   Security    DefaultAllocator::malloc(unsigned long)
Malloc 16 Bytes 2   < multiple >    32 Bytes    Security    DefaultAllocator::malloc(unsigned long)
Malloc 16 Bytes 1   0x620000014420  16 Bytes    Security    DefaultAllocator::malloc(unsigned long)
SecIdentity 1   0x628000267540  64 Bytes    Security    Security::SecCFObject::allocate(unsigned long, Security::CFClass const&)
```

This patch seems to take care of most of the leaks. There are two leaks that I can't seems to remediate.

```
Leaked Object   #   Address Size    Responsible Library Responsible Frame
Malloc 64 Bytes 1   0x60000087c180  64 Bytes    Security    DefaultAllocator::malloc(unsigned long)
Malloc 16 Bytes 1   0x60000001bde0  16 Bytes    Security    DefaultAllocator::malloc(unsigned long)
```

I have opened a bug with Apple. https://openradar.appspot.com/radar?id=5518388159840256

### CODE REVIEW

File: Source/MOLAuthenticatingURLSession.m [hunk 1]
```diff
@@ -242,7 +243,18 @@ - (NSURLCredential *)clientCredentialForProtectionSpace:(NSURLProtectionSpace *)
 ///  If the server's certificate chain does not evaluate for any reason, returns nil.
 ///
 - (NSURLCredential *)serverCredentialForProtectionSpace:(NSURLProtectionSpace *)protectionSpace {
-  SecTrustRef serverTrust = protectionSpace.serverTrust;
+  SecTrustRef serverTrust = NULL;
+  CFArrayRef policies = NULL;
```
Comment (russellhancox) [inline Source/MOLAuthenticatingURLSession.m]: Move lines 247-256 to line 289, just above where the anchor certs are set as we only need to do this if setting the anchor cert. Also, add a comment explaining why this is necessary.

**Comment by tburgin** [comment]
Done. I also added setters for the custom anchors.

File: Source/MOLAuthenticatingURLSession.m [hunk 2]
```diff
@@ -242,69 +271,71 @@ - (NSURLCredential *)clientCredentialForProtectionSpace:(NSURLProtectionSpace *)
...
       [self log:@"Server Trust: Could not set anchor certificates: %d", err];
+      if (serverTrust) CFRelease(serverTrust);
       return nil;
     }
   }
 
   // Print details about the server's leaf certificate.
-  SecCertificateRef firstCert = SecTrustGetCertificateAtIndex(serverTrust, 0);
+  SecCertificateRef firstCert =
+      SecTrustGetCertificateAtIndex(serverTrust ?: protectionSpace.serverTrust, 0);
```
Comment (russellhancox) [inline Source/MOLAuthenticatingURLSession.m]: You can avoid the elvis check here; given that the certs are copied from protectionSpace.serverTrust to serverTrust you can just use protectionSpace.serverTrust here.

File: Source/MOLAuthenticatingURLSession.m [hunk 3]
```diff
@@ -242,69 +271,71 @@ - (NSURLCredential *)clientCredentialForProtectionSpace:(NSURLProtectionSpace *)
...
-    NSString *pemStrings = [[NSString alloc] initWithData:self.serverRootsPemData
-                                                 encoding:NSASCIIStringEncoding];
-    NSArray *certs = [MOLCertificate certificatesFromPEM:pemStrings];
 
-    // Make a new array of the SecCertificateRef's from the MOLCertificate's.
-    NSMutableArray *certRefs = [[NSMutableArray alloc] initWithCapacity:certs.count];
-    for (MOLCertificate *cert in certs) {
-      [certRefs addObject:(id)cert.certRef];
-    }
+    // Create a copy of protectionSpace.serverTrust by grabbing it's policies and certificates.
```
Comment (russellhancox) [inline Source/MOLAuthenticatingURLSession.m]: nit: s/it's/its

File: Source/MOLAuthenticatingURLSession.m [hunk 4]
```diff
@@ -242,69 +271,71 @@ - (NSURLCredential *)clientCredentialForProtectionSpace:(NSURLProtectionSpace *)
...
+  // protectionSpace.serverTrust. If there are custom anchors serverTrust will be evaluated, if
+  // there are no custom anchors protectionSpace.serverTrust will be evaluated.
+  SecTrustRef serverTrust = NULL;
+
+  if (self.anchors) {
+    CFArrayRef policies = NULL;
+    SecTrustCopyPolicies(protectionSpace.serverTrust, &policies);
+    int certCount = (int)SecTrustGetCertificateCount(protectionSpace.serverTrust);
+    NSMutableArray *serverTrustCertRefs = [[NSMutableArray alloc] initWithCapacity:certCount];
+    for (int i = 0; i < certCount; i++) {
```
Comment (russellhancox) [inline Source/MOLAuthenticatingURLSession.m]: nit: s/i++/++i

**Comment by tburgin** [comment]
Done.

**Pull request merged**